jlm/llvm/opt/alias-analyses/Andersen.cpp

```diff
@@ -0,0 +1,702 @@
+
+  for (size_t n = 0; n < node.ninputs(); n++)
+  {
+    const auto & inputRegister = *node.input(n)->origin();
+    if (!is<PointerType>(inputRegister.type()))
+      continue;
+
+    // TODO: Pass pointer information through aggregate types
+    // Since the rest of the code only considers values of PointerType, mark inputs as escaping.
+    auto inputRegisterPO = Set_->GetRegisterPointerObject(inputRegister);
+    Constraints_->AddRegisterContentEscapedConstraint(inputRegisterPO);
+  }
+}
+
+void
+Andersen::AnalyzeConstantStruct(const rvsdg::simple_node & node)
+{
+  JLM_ASSERT(is<ConstantStruct>(&node));
+
+  for (size_t n = 0; n < node.ninputs(); n++)
+  {
+    const auto & inputRegister = *node.input(n)->origin();
+    if (!is<PointerType>(inputRegister.type()))
+      continue;
+
+    // TODO: Pass pointer information through aggregate types
+    // Since the rest of the code only considers values of PointerType, mark inputs as escaping.
+    const auto inputRegisterPO = Set_->GetRegisterPointerObject(inputRegister);
+    Constraints_->AddRegisterContentEscapedConstraint(inputRegisterPO);
+  }
+}
+
+void
+Andersen::AnalyzeConstantAggregateZero(const rvsdg::simple_node & node)
+{
+  JLM_ASSERT(is<ConstantAggregateZero>(&node));
+  auto & output = *node.output(0);
+
+  if (!is<PointerType>(output.type()))
+    return;
+
+  // ConstantAggregateZero cannot point to any memory location.
+  // We therefore only insert a register node for it, but let this node not point to anything.
+  Set_->CreateRegisterPointerObject(output);
+}
+
+void
+Andersen::AnalyzeExtractValue(const rvsdg::simple_node & node)
+{
+  JLM_ASSERT(is<ExtractValue>(&node));
+
+  const auto & result = *node.output(0);
+  if (!is<PointerType>(result.type()))
+    return;
+
+  // TODO: Make aggregate types with at least one field of pointer type have PointerObjects.
+  // Then we could be more precise than "escaping" all pointers passing through aggregate types.
+  // This involves replacing all usages of is<PointerType> with IsOrContains<PointerType>
+  const auto resultPO = Set_->CreateRegisterPointerObject(result);
+  Constraints_->AddPointsToExternalConstraint(resultPO);
+}
+
+void
+Andersen::AnalyzeStructuralNode(const rvsdg::structural_node & node)
+{
+  const auto & op = node.operation();
+  const auto & type = typeid(op);
+
+  if (type == typeid(lambda::operation))
+    AnalyzeLambda(*util::AssertedCast<const lambda::node>(&node));
+  else if (type == typeid(delta::operation))
+    AnalyzeDelta(*util::AssertedCast<const delta::node>(&node));
+  else if (type == typeid(phi::operation))
+    AnalyzePhi(*util::AssertedCast<const phi::node>(&node));
+  else if (type == typeid(rvsdg::gamma_op))
+    AnalyzeGamma(*util::AssertedCast<const rvsdg::gamma_node>(&node));
+  else if (type == typeid(rvsdg::theta_op))
+    AnalyzeTheta(*util::AssertedCast<const rvsdg::theta_node>(&node));
+  else
+    JLM_UNREACHABLE("Unknown structural node operation");
```

Comment:
I believe you should use `dynamic_cast<>()` on the node and then throw out the AssertedCast.